core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,108 +69,169 @@ class LogManager(val logDirs: Array[File],
 
+  private val dirLocks = lockLogDirs(liveLogDirs)
+
   // public, so we can access this from kafka.admin.DeleteTopicTest
   val cleaner: LogCleaner =
     if(cleanerConfig.enableCleaner)
-      new LogCleaner(cleanerConfig, logDirs, logs, time = time)
+      new LogCleaner(cleanerConfig, liveLogDirs, logs, time = time)
     else
       null
-  
+
+
   /**
    * Create and check validity of the given directories, specifically:
    * <ol>
    * <li> Ensure that there are no duplicates in the directory list
    * <li> Create each directory if it doesn't exist
-   * <li> Check that each path is a readable directory 
+   * <li> Check that each path is a readable directory
    * </ol>
    */
-  private def createAndValidateLogDirs(dirs: Seq[File]) {
+  private def createAndValidateLogDirs(dirs: Seq[File]): ArrayBuffer[File] = {
     if(dirs.map(_.getCanonicalPath).toSet.size < dirs.size)
-      throw new KafkaException("Duplicate log directory found: " + logDirs.mkString(", "))
-    for(dir <- dirs) {
-      if(!dir.exists) {
-        info("Log directory '" + dir.getAbsolutePath + "' not found, creating it.")
-        val created = dir.mkdirs()
-        if(!created)
-          throw new KafkaException("Failed to create data directory " + dir.getAbsolutePath)
+      throw new KafkaException("Duplicate log directory found: " + dirs.mkString(", "))
+
+    val liveLogDirs = ArrayBuffer.empty[File]
+
+    for (dir <- dirs) {
+      try {
+        if (!dir.exists) {
+          info("Log directory '" + dir.getAbsolutePath + "' not found, creating it.")
+          val created = dir.mkdirs()
+          if (!created)
+            throw new KafkaException("Failed to create data directory " + dir.getAbsolutePath)
+        }
+        if (!dir.isDirectory || !dir.canRead)
+          throw new KafkaException(dir.getAbsolutePath + " is not a readable log directory.")
+        liveLogDirs += dir
+      } catch {
+        case t: Throwable => error("Failed to create or validate data directory " + dir.getAbsolutePath, t);
       }
-      if(!dir.isDirectory || !dir.canRead)
-        throw new KafkaException(dir.getAbsolutePath + " is not a readable log directory.")
     }
+    if (liveLogDirs.isEmpty)
+      throw new KafkaException("None of the specified log dirs from " + dirs.mkString(", ") + " can be created or validated")
+
+    liveLogDirs
   }
-  
+
+  def handleLogDirFailure(dir: String) {
+    if (!logDirs.exists(_.getAbsolutePath == dir))
+      throw new RuntimeException(s"Log dir $dir is not found in the config.")
+
+    if (!liveLogDirs.exists(_.getAbsolutePath == dir)) {
+      info(s"Log dir $dir is already offline.")
+      return
+    }
+
+    logCreationOrDeletionLock synchronized {
+      if (!liveLogDirs.exists(_.getAbsolutePath == dir)) {
+        info(s"Log dir $dir is already offline.")
+        return
+      }
+      info(s"Stopping serving logs in dir $dir")
+
+      val newOfflineDirs = liveLogDirs.filter(_.getAbsolutePath == dir)
```

Comment:
Note that `liveLogDirs` has type `ArrayBuffer[File]` and `dir` has type `String`. It seems that `ArrayBuffer` only allows remove by position in the array. It does not have a method to remove element by value.